File: pastebin.py
import requests


def get_token(username, password):
    """
    Login to pastes.io and get API token

    Args:
        username (str): Username for the account
        password (str): Password for the account

    Returns:
        dict: Response from the API containing the API token and user information
    """
    url = "https://pastesio.com/api/login"

    headers = {
        'Accept': 'application/json'
    }

    payload = {
        'username': username,
        'password': password
    }

    files = []

    response = requests.post(url, data=payload, files=files, headers=headers)
    data = response.json()
    print(data)
    api_token = None
    if 'success' in data and 'api_token' in data['success']:
        api_token = data['success']['api_token']

    return api_token


def create_paste(token, title, content, status, syntax=None, expire=None, password=None):
    """
    Create a new paste on pastesio.com

    Args:
        token (str): Bearer Token (required)
        title (str): Title of the paste (max 80 chars)
        content (str): Content of the paste (required)
        status (int): Visibility status (required): 0=public, 1=unlisted, 2=private
        syntax (str, optional): Syntax highlighting slug
        expire (str, optional): Expiration time - options: N, 10M, 1H, 1D, 1W, 2W, 1M, 6M, 1Y, SD
        password (str, optional): Password protection (max 50 chars)

    Returns:
        dict: Response from the API
    """
    url = "https://pastesio.com/api/paste/create"

    headers = {
        'Accept': 'application/json',
        'Authorization': f'Bearer {token}'
    }

    # Prepare payload with required parameters
    payload = {
        'content': content,
        'status': status
    }

    # Add optional parameters if provided
    if syntax:
        payload['syntax'] = syntax
    if expire:
        payload['expire'] = expire
    if title:
        payload['title'] = title
    if password:
        payload['password'] = password

    response = requests.post(url, data=payload, headers=headers)
    return response.json()


def get_paste(token, slug, password=None):
    """
    Get a paste from pastesio.com by its slug

    Args:
        token (str): Bearer Token (required)
        slug (str): The paste's unique identifier
        password (str, optional): Password if the paste is password protected

    Returns:
        dict: Response from the API containing the paste
    """
    url = f"https://pastesio.com/api/pastes/{slug}"

    headers = {
        'Accept': 'application/json',
        'Authorization': f'Bearer {token}'
    }

    payload = {}
    if password:
        payload['password'] = password

    response = requests.post(url, data=payload, headers=headers)
    return response.json()


def update_paste(token, slug, content, title, status=None, syntax=None, password=None):
    """
    Update an existing paste on pastesio.com

    Args:
        token (str): Bearer Token (required)
        slug (str): The paste's unique identifier (required)
        title (str): Title of the paste (max 80 chars)
        content (str): Updated content of the paste (required)
        status (int, optional): Visibility status: 0=public, 1=unlisted, 2=private
        syntax (str, optional): Syntax highlighting slug
        password (str, optional): Password protection (max 50 chars)

    Returns:
        dict: Response from the API
    """
    url = "https://pastesio.com/api/paste/update"

    headers = {
        'Accept': 'application/json',
        'Authorization': f'Bearer {token}'
    }

    payload = {
        'slug': slug,
        'content': content
    }

    # Add optional parameters if provided
    if status is not None:
        payload['status'] = status
    if syntax:
        payload['syntax'] = syntax
    if title:
        payload['title'] = title
    if password:
        payload['password'] = password

    response = requests.post(url, data=payload, headers=headers)
    return response.json()


if __name__=="__main__":
    token = get_token(username="graveness638", password="r7I75YzJhbtxd!cnkeKRJ^eC")
    print(token)
    response = create_paste(token=token, content="testiiiing", title="test", status=2)
    print(response)

File: src/lib/services/pastebin.ts
const PASTE_TITLE = "ical_sync";
const PASTE_STATUS = 2;
const BASE_URL = "https://pastesio.com";

function proxify(url: string) {
    // return url;
    return `https://api.cors.lol/?url=${url}`;
    // return `https://corsproxy.io/?key=1f8a5538&url=${encodeURIComponent(url)}`;
    // return `https://corsproxy.io/?key=1f8a5538&url=${url}`;
}

async function getToken(username: string, password: string): Promise<string | null> {
    const url = proxify(`${BASE_URL}/api/login`);

    const headers = {
        Accept: "application/json",
    };

    const payload = {
        username: username,
        password: password,
    };

    try {
        const response = await fetch(url, {
            method: "POST",
            headers: headers,
            body: JSON.stringify(payload),
        });
        console.log(response);

        const data = await response.json();

        if (data.success && data.success.api_token) {
            return data.success.api_token;
        }

        return null;
    } catch (error) {
        console.error("Error getting token:", error);
        return null;
    }
}

// async function getPaste(token: string, slug: string, password?: string): Promise<any> {
//     const url = `${BASE_URL}/api/pastes/${slug}`;

//     const headers = {
//         Accept: "application/json",
//         Authorization: `Bearer ${token}`,
//         "Content-Type": "application/json",
//     };

//     const payload: Record<string, any> = {};
//     if (password) {
//         payload["password"] = password;
//     }

//     try {
//         const response = await fetch(url, {
//             method: "POST",
//             headers: headers,
//             body: Object.keys(payload).length ? JSON.stringify(payload) : undefined,
//         });

//         return await response.json();
//     } catch (error) {
//         console.error("Error fetching paste:", error);
//         throw error;
//     }
// }

async function createPaste(
    token: string,
    title: string,
    content: string,
    status: number,
    syntax?: string,
    expire?: string,
    password?: string,
): Promise<string | null> {
    const url = `${BASE_URL}/api/paste/create`;

    const headers = {
        Accept: "application/json",
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
    };

    const payload: Record<string, any> = {
        content: content,
        status: status,
    };

    if (syntax) {
        payload["syntax"] = syntax;
    }
    if (expire) {
        payload["expire"] = expire;
    }
    if (title) {
        payload["title"] = title;
    }
    if (password) {
        payload["password"] = password;
    }

    try {
        const response = await fetch(url, {
            method: "POST",
            headers: headers,
            body: JSON.stringify(payload),
        });
        console.log(response);

        const data = await response.json();

        if (data.success && data.success.slug) {
            return data.success.slug;
        }

        return null;
    } catch (error) {
        console.error("Error creating paste:", error);
        throw error;
    }
}

async function updatePaste(
    token: string,
    slug: string,
    content: string,
    title: string,
    status?: number,
    syntax?: string,
    password?: string,
): Promise<any> {
    const url = `${BASE_URL}/api/paste/update`;

    const headers = {
        Accept: "application/json",
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
    };

    const payload: Record<string, any> = {
        slug: slug,
        content: content,
    };

    if (status !== undefined) {
        payload["status"] = status;
    }
    if (syntax) {
        payload["syntax"] = syntax;
    }
    if (title) {
        payload["title"] = title;
    }
    if (password) {
        payload["password"] = password;
    }

    try {
        const response = await fetch(url, {
            method: "POST",
            headers: headers,
            body: JSON.stringify(payload),
        });
        console.log(response);

        return await response.json();
    } catch (error) {
        console.error("Error updating paste:", error);
        throw error;
    }
}

export async function storeData(data: string) {
    let username = localStorage.getItem("username");
    let password = localStorage.getItem("password");
    if (!username || !password) {
        throw Error("No username or password provided!");
    }
    let token = await getToken(username, password);
    if (!token) throw Error("Token generation failed.");
    let slug = localStorage.getItem("slug");
    if (!slug) {
        let response = await createPaste(token, PASTE_TITLE, data, PASTE_STATUS);
        if (!response) throw Error("Paste could not be created.");
    } else {
        let response = await updatePaste(token, slug, data, PASTE_TITLE);
        if (!response) throw Error("Paste could not be updated.");
    }
}

export function getPasteUrl(): string | null {
    let slug = localStorage.getItem("slug");
    if (!slug) return null;
    return `https://pastesio.com/raw/${slug}`;
}
